src/main/scala/tech/cryptonomic/conseil/util/RouteUtil.scala

```diff
@@ -0,0 +1,112 @@
+package tech.cryptonomic.conseil.util
+
+import java.util.UUID
+
+import akka.http.scaladsl.model.StatusCodes.InternalServerError
+import akka.http.scaladsl.model.{HttpEntity, HttpResponse, RemoteAddress, StatusCodes}
+import akka.http.scaladsl.server.Directives._
+import akka.http.scaladsl.server.directives.BasicDirectives
+import akka.http.scaladsl.server.{Directive, ExceptionHandler, RequestContext, Route}
+import akka.stream.Materializer
+import cats.effect.concurrent.MVar
+import cats.effect.{Concurrent, IO}
+import com.typesafe.scalalogging.LazyLogging
+import org.slf4j.MDC
+
+/** Utility class for recording responses */
+class RouteUtil(implicit concurrent: Concurrent[IO]) extends LazyLogging {
+
+  private val requestInfoMap: MVar[IO, Map[UUID, RequestValues]] =
+    MVar.of[IO, Map[UUID, RequestValues]](Map.empty[UUID, RequestValues]).unsafeRunSync()
+
+  /** Directive adding recorded values to the MDC */
+  def recordResponseValues(
+      ip: RemoteAddress
+  )(implicit materializer: Materializer, correlationId: UUID): Directive[Unit] =
+    BasicDirectives.extractRequestContext.flatMap { ctx =>
+      (for {
+        requestMap <- requestInfoMap.take
+        value = RequestValues.fromCtxAndIp(ctx, ip)
+        _ <- requestInfoMap.put(requestMap.updated(correlationId, value))
+      } yield ()).unsafeRunSync()
+
+      val response = BasicDirectives.mapResponse { resp =>
+        (for {
+          requestMap <- requestInfoMap.take
+          _ <- requestInfoMap.put(requestMap.filterNot(_._1 == correlationId))
+        } yield requestMap(correlationId).logResponse(resp)).unsafeRunAsyncAndForget()
+        resp
+      }
+      response
+    }
+
+  /** Custom exception handler with MDC logging */
+  def loggingExceptionHandler(implicit correlationId: UUID): ExceptionHandler =
+    ExceptionHandler {
+      case e: Throwable =>
+        val response = HttpResponse(InternalServerError)
+
+        (for {
+          requestMap <- requestInfoMap.take
+          _ <- requestInfoMap.put(requestMap.filterNot(_._1 == correlationId))
+        } yield requestMap(correlationId).logResponse(response)).unsafeRunAsyncAndForget()
```

Comment:
this piece could be extracted to a 
```scala
type RequestMap = Map[UUID, RequestValues]
def requestMapModify[A](modify: RequestMap => RequestMap)(useValues: RequestValues => A) = for {
  map <- requestInfoMap.take
  _   <- requestInfoMap.put(modify(map))
} yield useValues(map(correlationId))

val response = HttpResponse(InternalServerError)

requestMapModify(
  modify = _.filterNot(_._1 == correlationId)
){ values => 
  values.logResponse(response)
}.unsafeRunAsyncAndForget()

```
and so on... wdyt?

Reply:
sure, changed